WordPress/pattern-directory
# Directory: Add a system for flagging patterns

This adds the REST API endpoint, flag schema, and back-end UI described in #61. The front end UI will probably get a separate PR, and maybe a separate issue as well.

This introduces the `wporg-pattern-flag` post type, the `wporg-pattern-flag-reason` taxonomy, and the `resolved` post status to store and manage data when a pattern in the pattern directory is flagged for review. A flag is comprised of:

* A reason for the flag, stored as a term in the `wporg-pattern-flag-reason` taxonomy. There will be a few pre-defined terms that a user can choose from on the front end as the reason for the flag.
* Details about the flag, stored in `post_excerpt`. This provides more context for the moderator reviewing the pattern and its flags.
* The ID of the "parent" pattern post, stored in `post_parent`
* The ID of the user who submitted the flag, stored in `post_author`.
* The status of the flag, stored in `post_status`. A new flag starts as `pending` and can be marked as `resolved` once it has been reviewed.

The still-in-the-works front end form will submit the flag data to the REST API endpoint, which is a slightly modified version of the standard endpoint for WP post types. The main differences are:

* To view flags, you must have edit permissions for the `wporg-pattern` post type.
* To create flags, you don't need edit permissions, but you must be logged into your wp.org account.
* There are checks to prevent submitting a flag to a pattern that isn't published, and to prevent submitting a flag to a pattern that you've already flagged previously, regardless of the status of previous flags.

For the back-end UI, this PR focuses on screens directly related to the new post type and taxonomy, namely a list table and a UI for creating new terms. Adding info about a pattern's flags to the patterns list table will be covered in a separate issue/PR.

The list table is a standard posts list table that has been modified via filters and actions. Of note are the custom columns, row actions, and bulk actions to facilitate reviewing flags.

### Screenshot

![Screen Shot 2021-04-16 at 11 50 03](https://user-images.githubusercontent.com/916023/114922649-c64b7080-9de0-11eb-82d8-f703fce00c0e.png)

### How to test the changes in this Pull Request:

**Setup**

* Create one or more block patterns.
* Create one or more terms on the Reasons subpage.

**Test the API**

* This is most easily done with a client app like [Insomnia](https://insomnia.rest/).
* To create a new flag, send a POST request to `http://localhost:8888/wp-json/wp/v2/wporg-pattern-flag`. It needs to have the following parameters:
  * `parent`: The post ID of a pattern post
  * `wporg-pattern-flag-reason`: The ID of a Reason term
  * `excerpt`: A "details" string
* To get existing flags, send a GET request to `http://localhost:8888/wp-json/wp/v2/wporg-pattern-flag` with the usual optional parameters for any kind of posts.

**Test the back end UI**

* Once you have some flags created, point your browser to `http://localhost:8888/wp-admin/edit.php?post_type=wporg-pattern-flag`.
* Try resolving/unresolving and trashing/untrashing flags with row actions (links that appear when you hover over a flag row), and bulk actions.
* Try viewing only flags for one particular pattern.
